src/server/scripts/EasternKingdoms/ScarletEnclave/chapter1.cpp

```diff
@@ -883,6 +884,60 @@ class npc_ros_dark_rider : public CreatureScript
 
 };
 
+enum GiftOfTheHarvester
+{
+    SPELL_GIFT_OF_THE_HARVESTER = 52479,
+    SPELL_MINER_GHOUL_TRANSFORM = 52490,
+    SPELL_MINER_GHOST_TRANSFORM = 52505
+};
+
+struct gift_of_the_harvesterAI : public GameObjectAI
+{
+    gift_of_the_harvesterAI(GameObject* go) : GameObjectAI(go) { }
+
+    void Reset() override
+    {
+        _scheduler.Schedule(2s, [this](TaskContext /*context*/)
+        {
+            me->SendCustomAnim(0);
+            me->CastSpell(nullptr, SPELL_GIFT_OF_THE_HARVESTER, true);
+        });
+    }
+    
+    void UpdateAI(uint32 diff) override
+    {
+        _scheduler.Update(diff);
+    }
+
+private:
+    TaskScheduler _scheduler;
+};
+
+struct spell_gift_of_the_harvester : public SpellScript
+{
+    PrepareSpellScript(spell_gift_of_the_harvester);
+
+    bool Validate(SpellInfo const* /*spell*/) override
+    {
+        return ValidateSpellInfo(
+                {
+                    SPELL_MINER_GHOUL_TRANSFORM,
+                    SPELL_MINER_GHOST_TRANSFORM
+                });
+    }
+
+    void HandleScriptEffect(SpellEffIndex /*effIndex*/)
+    {
+        uint32 spellId = RAND(SPELL_MINER_GHOUL_TRANSFORM, SPELL_MINER_GHOST_TRANSFORM);
+        GetCaster()->CastSpell(GetHitUnit(), spellId, true);
```

Comment:
GetCaster() returns Unit* so ofc it will return NULL when the caster is a GameObject
https://github.com/TrinityCore/TrinityCore/blob/a926a3088d39897fe3ab58239775661449498363/src/server/game/Spells/SpellScript.h#L379

Try GetGObjCaster() instead (and NULL check it too)
https://github.com/TrinityCore/TrinityCore/blob/a926a3088d39897fe3ab58239775661449498363/src/server/game/Spells/SpellScript.h#L860

After that set a breakpoint here and debug it to ensure everything works as expected

Reply:
Works like a charm now. Thanks a lot!